File: proposed-projects/round-trip-experiment/reverse/20260909T185402Z_14873f98/20260910T102153Z_a4bbfe64/ported_implementation/src/grammarGraph/getParentStackId.ts
import { Color } from './colorize';
import type { Colorize } from './colorize';
import type { GraphPointer } from './graphPointer';

export const getParentStackId = (
  pointer: GraphPointer,
  col: Colorize,
): string => {
  const stackIds: string[] = [];
  let parent = pointer.parent;
  while (parent !== undefined) {
    const { stackId, pathId, stepId } = parent.node.meta;
    stackIds.push(`${stackId},${pathId},${stepId}`);
    parent = parent.parent;
  }
  return stackIds.map((i) => col(i, Color.RED)).join(col('<-', Color.GRAY));
};

File: packages/gbnf-experiment/data/20260909T194429Z_a7bb3ecf/ported_implementation/grammar_graph/get_parent_stack_id.py
from __future__ import annotations

from typing import TYPE_CHECKING, List

from .colorize import Color, Colorize

if TYPE_CHECKING:  # pragma: no cover - typing only
    from .graph_pointer import GraphPointer


def get_parent_stack_id(pointer: 'GraphPointer', col: Colorize) -> str:
    stack_ids: List[str] = []
    parent = pointer.parent
    while parent:
        meta = parent.node.meta
        stack_ids.append(f'{meta.stack_id},{meta.path_id},{meta.step_id}')
        parent = parent.parent
    return col('<-', Color.GRAY).join(col(id_, Color.RED) for id_ in stack_ids)
